src/types/utils.cpp

```diff
@@ -428,6 +428,90 @@ catch (...)
     return {};
 }
 
+// Routine Description:
+// - Pre-process text pasted (presumably from the clipboard) with provided option.
+// Arguments:
+// - wstr - String to process.
+// - option - option to use.
+// Return Value:
+// - The result string.
+std::wstring Utils::FilterStringForPaste(const std::wstring_view wstr, const FilterOption option)
+{
+    std::wstring filtered;
+    filtered.reserve(wstr.length());
+
+    const auto isControlCode = [](wchar_t c) {
+        if (c >= L'\x20' && c <= L'\x7f')
+        {
+            // Printable ASCII characters + DEL.
+            return false;
+        }
+
+        if (c > L'\x7f')
+        {
+            // Not a control code.
+            return false;
+        }
+
+        // All ASCII control codes will be removed except HT(0x09), LF(0x0a),
+        // CR(0x0d) and DEL(0x7F).
+        return c >= L'\x00' && c <= L'\x08' ||
+               c >= L'\x0b' && c <= L'\x0c' ||
+               c >= L'\x0e' && c <= L'\x1f';
+    };
+
+    std::wstring::size_type pos = 0;
+    std::wstring::size_type begin = 0;
+
+    while (pos < wstr.size())
+    {
+        const wchar_t c = wstr.at(pos);
+
+        if (WI_IsFlagSet(option, FilterOption::CarriageReturnNewline) && c == L'\n')
+        {
+            // copy up to but not including the \n
+            filtered.append(wstr.cbegin() + begin, wstr.cbegin() + pos);
+            if (!(pos > 0 && (wstr.at(pos - 1) == L'\r')))
+            {
+                // there was no \r before the \n we did not copy,
+                // so append our own \r (this effectively replaces the \n
+                // with a \r)
+                filtered.push_back(L'\r');
+            }
+            ++pos;
+            begin = pos;
+        }
+        else if (WI_IsFlagSet(option, FilterOption::ControlCodes) && isControlCode(c))
+        {
+            // copy up to but not including the control code
+            filtered.append(wstr.cbegin() + begin, wstr.cbegin() + pos);
+            ++pos;
+            begin = pos;
+        }
+        else
+        {
+            ++pos;
+        }
+    }
+
+    // If we entered the while loop even once, begin would be non-zero
+    // (because we set begin = pos right after incrementing pos)
+    // So, if begin is still zero at this point it means we never found a newline
+    // and we can just write the original string
+    if (begin == 0)
+    {
+        return std::wstring(wstr);

```

Comment:
Does this still copy the `wstr`? I have to return a `std::wstring` here. Don't know if there's a better way.

Reply:
I think it's okay to copy it here. We can look later at how to remove this last copy.

Prefer `std::wstring{ wstr }` or just `return wstr;` (if there is an automatic converter)

Reply:
Sadly no automatic converter available.